### DISCUSSION TITLE
Updating to Csla 7 - DataPortal.DataPortalInvoke

### DISCUSSION THREAD

**Discussion by ajohnstone-ks** [MarimerLLC/csla#3787] [Questions] [answered]
I didn't see any questions about DataPortalInvoke, the static event that existing in prior to Csla 6 I guess.  Is it just gone?  Is there an alternate method to hook into for every data portal call?

**Comment by rockfordlhotka** [comment, marked as answer]
The new data portal is instanced, not static, and so there is no central way to do that sort of thing using events. At least not a safe way in terms of memory leaks.

On the _server_ side the interceptor concept provides a rich way to be part of the processing pipeline. On the _client_ side, the best way to hook into all data portal calls is to create an use a custom proxy (subclass of whatever proxy you are otherwise using).

**Comment by rockfordlhotka** [comment]
I suppose that creating a proxy subclass isn't terribly elegant. Perhaps we should consider a client-side interceptor model like we have on the server? Presumably this is primarily for logging?

**Reply by StefanOssendorf** [reply]
I can recommend a Decorator approach on the client side. We are using [Scrutor](https://github.com/khellang/Scrutor) on the client side to hook into everything.
And the great thingb is you can create as many decorators as you want to specialize them.

**Reply by ajohnstone-ks** [reply]
@rockfordlhotka In our app when  user logs in we build a session object that has all their permissions (this is way before claims).  We return that object to the client with a secure RNG key.  When requests come to the server, we use that key to avoid reloading the session, which is expensive.  But we only store this session data in the web server memory (also not great, but I can't immediately address it).  If the session isn't found on the web server, which happens due to recycling, then we silently re-login the user.  

Although a client side interceptor would be nice in general, I think our particular problem is very specific to us.  I think I have come up with a workaround, but it will be some time until we have a code base that can be compiled & run again, so I don't know for sure how well it will work.

**Reply by StefanOssendorf** [reply]
Assuming you use http as your transport protocol. That sounds like you should use custom header to send the key back to the server. It's effectively acting like an access token.

**Reply by ajohnstone-ks** [reply]
@StefanOssendorf Yes, its an access token.  The issue is not getting the token from client to server, its the fact that if the token isn't found on the server, the user is effectively logged out, so we handle that client side by logging them back in.  That was previously happening in the DPI handler, but since that's gone I need to go another route.  Ideally our server side session data would be out of process, but that's something we haven't been able to tackle.

**Reply by StefanOssendorf** [reply]
It depends on the stage where you can/want to log the user back in. With ASP.Net core you have multiple ways:
* Middleware which does the magic if CSLA is not involved
* As rocky mentioned implement and register an interceptor

**Reply by ajohnstone-ks** [reply]
@StefanOssendorf Right, but as Rocky mentioned there is no client-side interceptor.  The login needs to happen from the client side.  But we do already have a subclass of HttpProxy, so I can change that to handle this for now.
